File: scraper.js
// Lookout v2.1 - Nitter Scraper with Playwright
const { chromium } = require('playwright');
const { WebhookClient, EmbedBuilder } = require('discord.js');
const fs = require('fs').promises;
const path = require('path');
const { getCachedHealthyInstances } = require('./healthChecker');
require('dotenv').config();

// ============================================================================
// CONFIGURATION  
// ============================================================================

// Nitter instances to try (priority order)
const NITTER_INSTANCES = [
  'nitter.net',            // Original, currently working
  'xcancel.com',
  'nitter.poast.org',
  'nitter.privacydev.net',
  'nitter.catsarch.com',
  'nitter.woodland.cafe',
  'nitter.in.projectsegfau.lt',
  'nitter.1d4.us',
  'nitter.moomoo.me',
  'nitter.eu.projectsegfau.lt',
  'nitter.projectsegfau.lt',
  'nitter.fdn.fr',
  'nitter.ktachibana.party',
  'nitter.pw'
];

const PLAYWRIGHT_TIMEOUT = 30000;  // 30 seconds
const HEALTH_CHECK_COUNT = 3;      // Check top 3 instances

// ============================================================================
// SCRAPING FUNCTION
// ============================================================================

async function scrapeNitterInstance(instance, handle) {
  console.log(`\n🔍 Trying ${instance}...`);

  const browser = await chromium.launch({
    headless: true,
    args: ['--no-sandbox', '--disable-setuid-sandbox']
  });

  try {
    const context = await browser.newContext({
      userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
    });

    const page = await context.newPage();
    const url = `https://${instance}/${handle}`;

    console.log(`   Navigating to ${url}...`);
    await page.goto(url, {
      timeout: PLAYWRIGHT_TIMEOUT,
      waitUntil: 'networkidle'
    });

    // Wait a moment for dynamic content
    await page.waitForTimeout(2000);

    console.log(`   Looking for tweets...`);

    // Find all tweet status links
    const statusLinks = await page.$$eval('a[href*="/status/"]', links =>
      links.map(link => link.href || link.getAttribute('href'))
    );

    if (statusLinks.length === 0) {
      throw new Error('No tweet links found on page');
    }

    console.log(`   Found ${statusLinks.length} tweet links`);

    // Get first tweet URL (skip checking for pinned - just get the latest)
    // We'll handle pinned tweets by checking the URL against our log
    const firstTweetLink = statusLinks[0];
    const tweetUrl = firstTweetLink.startsWith('http')
      ? firstTweetLink
      : `https://twitter.com${firstTweetLink}`;

    console.log(`   Latest tweet: ${tweetUrl}`);

    // Extract author info from page
    const authorName = await page.$eval(
      'a.fullname, .fullname',
      el => el.textContent.trim()
    ).catch(() => 'Unknown');

    const authorHandle = await page.$eval(
      'a.username, .username',
      el => el.textContent.trim()
    ).catch(() => '@unknown');

    // Get profile pic
    const profilePicUrl = await page.$eval(
      'img.avatar, .avatar img',
      el => el.src || el.getAttribute('src')
    ).catch(() => '');

    // Get tweet text (first one on page)
    const tweetText = await page.$eval(
      '.tweet-content',
      el => el.textContent.trim()
    ).catch(() => 'Tweet content unavailable');

    // Get timestamp (use current time if Nitter doesn't provide valid one)
    let timestamp = new Date().toISOString();
    try {
      const timeAttr = await page.$eval(
        '.tweet-date a, a[title]',
        el => el.getAttribute('title')
      );
      // Validate it's a real date before using it
      if (timeAttr && !isNaN(Date.parse(timeAttr))) {
        timestamp = timeAttr;
      }
    } catch (e) {
      // Use default current time
    }

    // Get image if present
    const tweetImageUrl = await page.$eval(
      '.still-image, .attachment img',
      el => el.src || el.getAttribute('src')
    ).catch(() => null);

    // Fix relative URLs for profile pic and images
    const fixedProfilePic = profilePicUrl && profilePicUrl.startsWith('/')
      ? `https://${instance}${profilePicUrl}`
      : profilePicUrl;

    const fixedImageUrl = tweetImageUrl && tweetImageUrl.startsWith('/')
      ? `https://${instance}${tweetImageUrl}`
      : tweetImageUrl;

    console.log(`✅ Successfully scraped from ${instance}!`);

    return {
      tweetUrl,
      authorName,
      authorHandle,
      tweetText,
      timestamp,
      profilePicUrl: fixedProfilePic,
      tweetImageUrl: fixedImageUrl,
      handle,
      nitterInstance: instance
    };

  } finally {
    await browser.close();
  }
}

// ============================================================================
// INSTANCE SELECTION & FALLBACK
// ============================================================================

async function fetchLatestTweet(handle) {
  // Get healthy instances
  const healthyInstances = await getCachedHealthyInstances(NITTER_INSTANCES, HEALTH_CHECK_COUNT);

  // Prioritize healthy instances, then try all
  const healthyDomains = healthyInstances.map(h => h.instance);
  const fallbackInstances = NITTER_INSTANCES.filter(i => !healthyDomains.includes(i));
  const instancesToTry = [...healthyDomains, ...fallbackInstances];

  console.log(`\n🎯 Will try ${instancesToTry.length} instances (${healthyDomains.length} pre-checked healthy)\n`);

  let lastError = null;

  for (const instance of instancesToTry) {
    try {
      const tweetData = await scrapeNitterInstance(instance, handle);
      return tweetData;
    } catch (error) {
      console.log(`   ❌ ${instance} failed: ${error.message}`);
      lastError = error;
    }
  }

  throw new Error(`All ${instancesToTry.length} Nitter instances failed. Last error: ${lastError?.message}`);
}

// ============================================================================
// DISCORD WEBHOOK
// ============================================================================

async function sendToDiscord(tweetData) {
  const webhookUrl = process.env.DISCORD_WEBHOOK_URL;
  if (!webhookUrl) {
    console.error('Error: DISCORD_WEBHOOK_URL is not set in the .env file.');
    return;
  }

  const webhookClient = new WebhookClient({ url: webhookUrl });
  const embed = new EmbedBuilder()
    .setColor(0x1DA1F2)
    .setAuthor({
      name: `${tweetData.authorName} • ${tweetData.authorHandle}`,
      iconURL: tweetData.profilePicUrl,
      url: tweetData.tweetUrl
    })
    .setDescription(tweetData.tweetText)
    .setTimestamp(new Date(tweetData.timestamp))
    .setFooter({ text: `🔭 Lookout via ${tweetData.nitterInstance}` });

  if (tweetData.tweetImageUrl) {
    embed.setImage(tweetData.tweetImageUrl);
  }

  console.log(`\n📤 Sending tweet from @${tweetData.handle} to Discord...`);
  try {
    await webhookClient.send({
      username: 'Lookout',
      avatarURL: tweetData.profilePicUrl,
      embeds: [embed],
    });
    console.log('✅ Successfully sent to Discord!');
  } catch (error) {
    console.error('❌ Error sending to Discord:', error.message);
    throw error;
  }
}

// ============================================================================
// MAIN
// ============================================================================

async function main() {
  const handle = process.argv[2];
  if (!handle) {
    console.error('❌ Error: No Twitter handle provided. Usage: node scraper.js <handle>');
    return;
  }

  console.log(`\n${'='.repeat(60)}`);
  console.log(`🔭 LOOKOUT v2.1 - Monitoring @${handle}`);
  console.log(`${'='.repeat(60)}\n`);

  const logFilePath = path.join(__dirname, 'log', `${handle}_last_tweet.txt`);

  // Read last posted tweet URL
  let lastPostedUrl = '';
  try {
    lastPostedUrl = await fs.readFile(logFilePath, 'utf-8');
    console.log(`📝 Last posted URL: ${lastPostedUrl}`);
  } catch (error) {
    console.log('📝 No previous log found. This is the first run.');
  }

  try {
    // Fetch latest tweet from Nitter
    const tweetData = await fetchLatestTweet(handle);

    // Check if it's a new tweet
    if (tweetData.tweetUrl === lastPostedUrl) {
      console.log('\n✓ No new tweet found. Exiting.');
      return;
    }

    console.log('\n' + '='.repeat(60));
    console.log('🆕 NEW TWEET DETECTED!');
    console.log('='.repeat(60));
    console.log(`👤 Author: ${tweetData.authorName} (${tweetData.authorHandle})`);
    console.log(`📝 Text: ${tweetData.tweetText.substring(0, 200)}${tweetData.tweetText.length > 200 ? '...' : ''}`);
    console.log(`🔗 URL: ${tweetData.tweetUrl}`);
    console.log(`🖼️  Image: ${tweetData.tweetImageUrl ? 'Yes' : 'No'}`);
    console.log('='.repeat(60));

    // Send to Discord
    await sendToDiscord(tweetData);

    // Update log file
    console.log(`\n💾 Updating log file...`);
    await fs.mkdir(path.dirname(logFilePath), { recursive: true });
    await fs.writeFile(logFilePath, tweetData.tweetUrl);

    console.log('\n✅ DONE!\n');

  } catch (error) {
    console.error(`\n${'='.repeat(60)}`);
    console.error(`❌ ERROR for @${handle}`);
    console.error(`${'='.repeat(60)}`);
    console.error(error.message);
    console.error('='.repeat(60) + '\n');
    process.exit(1);
  }
}

main();

File: scraper.py
#!/usr/bin/env python3
"""
Lookout v3.0 - Twitter Monitor using ntscraper
Simplified, faster, more reliable than Playwright approach
"""

import os
import sys
import json
from pathlib import Path
from ntscraper import Nitter
import requests
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

def send_to_discord(webhook_url, tweet_data):
    """Send tweet to Discord webhook"""
    
    embed = {
        "color": 0x1DA1F2,  # Twitter blue
        "author": {
            "name": f"{tweet_data['author']} • @{tweet_data['handle']}",
            "url": tweet_data['link']
        },
        "description": tweet_data['text'],
        "timestamp": tweet_data.get('date', ''),
        "footer": {
            "text": "🔭 Lookout v3.0"
        }
    }
    
    # Add image if present
    if tweet_data.get('pictures'):
        embed["image"] = {"url": tweet_data['pictures'][0]}
    
    payload = {
        "username": "Lookout",
        "embeds": [embed]
    }
    
    response = requests.post(webhook_url, json=payload)
    response.raise_for_status()
    return response.status_code

def get_latest_tweet(handle):
    """Fetch latest tweet using ntscraper"""
    print(f"\n{'='*60}")
    print(f"🔭 LOOKOUT v3.0 - Monitoring @{handle}")
    print('='*60)
    
    scraper = Nitter()
    
    print(f"\n🔍 Fetching latest tweet from @{handle}...")
    
    try:
        tweets = scraper.get_tweets(handle, mode='user', number=1)
        
        if not tweets or 'tweets' not in tweets or not tweets['tweets']:
            raise Exception("No tweets found")
        
        latest = tweets['tweets'][0]
        
        print(f"✅ Found tweet!")
        print(f"   Text: {latest['text'][:60]}...")
        print(f"   Link: {latest['link']}")
        
        return latest
        
    except Exception as e:
        print(f"❌ Error fetching tweet: {e}")
        raise

def main():
    if len(sys.argv) < 2:
        print("❌ Error: No Twitter handle provided")
        print("Usage: python scraper.py <handle>")
        sys.exit(1)
    
    handle = sys.argv[1]
    
    # Setup paths
    log_dir = Path("log")
    log_dir.mkdir(exist_ok=True)
    log_file = log_dir / f"{handle}_last_tweet.txt"
    
    # Read last posted URL
    last_posted_url = ""
    if log_file.exists():
        last_posted_url = log_file.read_text().strip()
        print(f"📝 Last posted URL: {last_posted_url}")
    else:
        print("📝 No previous log found. This is the first run.")
    
    try:
        # Fetch latest tweet
        tweet = get_latest_tweet(handle)
        tweet_url = tweet['link']
        
        # Check if it's new
        if tweet_url == last_posted_url:
            print("\n✓ No new tweet found. Exiting.")
            return
        
        print(f"\n{'='*60}")
        print("🆕 NEW TWEET DETECTED!")
        print('='*60)
        print(f"👤 Author: {tweet.get('author', 'Unknown')}")
        print(f"📝 Text: {tweet['text'][:200]}{'...' if len(tweet['text']) > 200 else ''}")
        print(f"🔗 URL: {tweet_url}")
        print(f"🖼️  Images: {len(tweet.get('pictures', []))}")
        print('='*60)
        
        # Send to Discord
        webhook_url = os.getenv('DISCORD_WEBHOOK_URL')
        if not webhook_url:
            print("\n⚠️  Warning: DISCORD_WEBHOOK_URL not set in .env file")
            print("Skipping Discord notification.")
        else:
            print(f"\n📤 Sending tweet to Discord...")
            status = send_to_discord(webhook_url, tweet)
            print(f"✅ Successfully sent to Discord! (HTTP {status})")
        
        # Update log file
        print(f"\n💾 Updating log file...")
        log_file.write_text(tweet_url)
        
        print("\n✅ DONE!\n")
        
    except Exception as e:
        print(f"\n{'='*60}")
        print(f"❌ ERROR for @{handle}")
        print('='*60)
        print(str(e))
        print('='*60 + '\n')
        sys.exit(1)

if __name__ == "__main__":
    main()
